src/main.rs

```diff
@@ -1,46 +1,39 @@
+mod argument_parser;
+mod linking;
+
 use std::{
     borrow::Cow,
     env,
     fs::{self, File},
     io::Write,
     ops::RangeInclusive,
     path::{Path, PathBuf},
-    process::{self, Command},
+    process,
 };
 
 use anyhow::anyhow;
 use object::{elf, Object as _, ObjectSection, SectionFlags};
 
 const EXIT_CODE_FAILURE: i32 = 1;
-// TODO make this configurable (via command-line flag or similar)
-const LINKER: &str = "rust-lld";
 /// Stack Pointer alignment required by the ARM architecture
 const SP_ALIGN: u64 = 8;
 
 fn main() -> anyhow::Result<()> {
-    notmain().map(|code| process::exit(code))
-}
-
-fn notmain() -> anyhow::Result<i32> {
     env_logger::init();
 
     // NOTE `skip` the name/path of the binary (first argument)
     let args = env::args().skip(1).collect::<Vec<_>>();
 
-    // link normally
-    let mut c1 = Command::new(LINKER);
-    c1.args(&args);
-    log::trace!("{:?}", c1);
-    let status = c1.status()?;
-    if !status.success() {
-        return Ok(status.code().unwrap_or(EXIT_CODE_FAILURE));
+    {
+        let exit_status = linking::link_normally(&args)?;
+        if !exit_status.success() {
+            process::exit(exit_status.code().unwrap_or(EXIT_CODE_FAILURE))
```

Comment:
don't call `process::exit` from this function. `exit` is divergent and terminates the program without `drop`-ing (destroying) variables on the caller (and parent) frames; this can result in temporary directories / files not being cleaned up (if the API you are using relies on destructors doing cleanup). I would revert this to call `exit` from the a top-level `main` function that has no stack variables

Reply:
I see. Then this is probably the main reason for the `fn notmain` pattern, then?